File: frontend/src/stores/networks.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import api from '@/composables/useDockerAPI'

export const useNetworksStore = defineStore('networks', () => {
  const networks = ref([])
  const loading = ref(false)
  const error = ref(null)

  async function fetchNetworks() {
    loading.value = true
    error.value = null
    try {
      const response = await api.get('/networks')
      networks.value = response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  async function inspectNetwork(id) {
    try {
      const response = await api.get(`/networks/${id}`)
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  async function createNetwork(config) {
    try {
      const response = await api.post('/networks', config)
      await fetchNetworks()
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  async function removeNetwork(id) {
    try {
      await api.delete(`/networks/${id}`)
      await fetchNetworks()
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  async function connectContainer(networkId, containerId) {
    try {
      const response = await api.post(`/networks/${networkId}/connect`, { container_id: containerId })
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  async function disconnectContainer(networkId, containerId) {
    try {
      const response = await api.post(`/networks/${networkId}/disconnect`, { container_id: containerId })
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  async function pruneNetworks() {
    try {
      const response = await api.post('/networks/prune')
      await fetchNetworks()
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  return {
    networks,
    loading,
    error,
    fetchNetworks,
    inspectNetwork,
    createNetwork,
    removeNetwork,
    connectContainer,
    disconnectContainer,
    pruneNetworks
  }
})

File: backend/api/handlers/networks.go
package handlers

import (
	"encoding/json"
	"net/http"
	"sunspear/services"

	"github.com/gorilla/mux"
)

type NetworkHandler struct {
	dockerService *services.DockerService
}

func NewNetworkHandler(dockerService *services.DockerService) *NetworkHandler {
	return &NetworkHandler{dockerService: dockerService}
}

func (h *NetworkHandler) ListNetworks(w http.ResponseWriter, r *http.Request) {
	networks, err := h.dockerService.ListNetworks(r.Context())
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondJSON(w, http.StatusOK, networks)
}

func (h *NetworkHandler) CreateNetwork(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Name     string `json:"name"`
		Driver   string `json:"driver"`
		Internal bool   `json:"internal"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	if req.Name == "" {
		http.Error(w, "name is required", http.StatusBadRequest)
		return
	}

	if req.Driver == "" {
		req.Driver = "bridge"
	}

	network, err := h.dockerService.CreateNetwork(r.Context(), req.Name, req.Driver, req.Internal)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondJSON(w, http.StatusCreated, network)
}

func (h *NetworkHandler) InspectNetwork(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id := vars["id"]

	network, err := h.dockerService.InspectNetwork(r.Context(), id)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}

	respondJSON(w, http.StatusOK, network)
}

func (h *NetworkHandler) RemoveNetwork(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id := vars["id"]

	if err := h.dockerService.RemoveNetwork(r.Context(), id); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondJSON(w, http.StatusOK, map[string]string{"status": "removed"})
}

func (h *NetworkHandler) ConnectContainer(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	networkID := vars["id"]

	var req struct {
		ContainerID string `json:"containerId"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	if req.ContainerID == "" {
		http.Error(w, "containerId is required", http.StatusBadRequest)
		return
	}

	if err := h.dockerService.ConnectNetwork(r.Context(), networkID, req.ContainerID); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondJSON(w, http.StatusOK, map[string]string{"status": "connected"})
}

func (h *NetworkHandler) DisconnectContainer(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	networkID := vars["id"]

	var req struct {
		ContainerID string `json:"containerId"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	if req.ContainerID == "" {
		http.Error(w, "containerId is required", http.StatusBadRequest)
		return
	}

	if err := h.dockerService.DisconnectNetwork(r.Context(), networkID, req.ContainerID); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	respondJSON(w, http.StatusOK, map[string]string{"status": "disconnected"})
}

func (h *NetworkHandler) PruneNetworks(w http.ResponseWriter, r *http.Request) {
	report, err := h.dockerService.PruneNetworks(r.Context())
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	response := map[string]interface{}{
		"deleted": report.NetworksDeleted,
	}

	respondJSON(w, http.StatusOK, response)
}
